### ISSUE TITLE
On Memorization in Probabilistic Deep Generative Models

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#145]
## TL; DR

- Propose Leave-One-Out based formulation for measuring probabilistic generative model's dependency on a single data instance. (=memorization)

## Motivation

- Memorization is different from overfitting; memorization is local dependency on single (or few) samples for a model's density estimation.
- Under memorization, a generative model's samples from the corresponding density region will likely result in redundant / limited outputs.

## Prior Work

- Approaches to evaluate memorization in supervised learning setting.
- Remotely related, but a recent paper on GAN replication (ICCV'21).

## Method

![image](https://user-images.githubusercontent.com/46308407/147901041-df472bb0-21cb-4c21-b24d-663a3c32f794.png)

- With memorization, density estimation heavily depends on a single example, at least locally (*dark blue*).

![image](https://user-images.githubusercontent.com/46308407/147901076-5513847d-aa7c-48f5-bffb-3525eb5d5fde.png)

- Model agnostic form of density estimation. *a* refers to a particular model, and *D* stands for the training dataset.

![image](https://user-images.githubusercontent.com/46308407/147901123-352124e9-b17b-4f6c-bbe9-91779fe1ab9b.png)

- Leave-One-Out memorization score for a data sample (indexed by *i*).

![image](https://user-images.githubusercontent.com/46308407/147901143-f6ade9ec-505f-4288-b405-372cb0539c10.png)

- Practical approximation of instance memorization score based on K-fold splits.
- Essence is, compute likelihood from models that have ***seen*** the specific sample, and from that subtract predicted likelihood of models that have ***not seen*** the sample.

![image](https://user-images.githubusercontent.com/46308407/147901178-e87a9bad-3dc3-4541-8f84-eb8fe23b6e7c.png)

## Experiments

- Experiments done on toy datasets: MNIST, CIFAR-10, and CelebA.

### Memorizations are sometimes counterintuitive

![image](https://user-images.githubusercontent.com/46308407/147901210-8fd78e49-3854-4e61-83a3-c8c3c3288a8e.png)

- Samples with high memorization scores are not always visually anomalous.

### Memorized samples are not always allocated low density

![image](https://user-images.githubusercontent.com/46308407/147901245-c32dd543-0c55-4062-b0dc-bc59e35d832d.png)

- It is true that there are more highly-memorized samples in low density region, still more than half of highly-memorized samples (quantile > 0.95) reside in central 90 quantile interval.

### Memorization can happen early

![image](https://user-images.githubusercontent.com/46308407/147901294-83e3ef45-3c53-4bef-8e9c-8edfd5493871.png)

- This shows that in relatively early epochs where generalization gap is yet to clearly form, we can observe quite high values of memorization scores. 
- This indicates that local memorization is different from global overfitting.
- Also, memorization score can be employed to assess different model architectures and hyperparameter configurations. (learning rate in this figure)

### Nearest neighbor test is not omnipotent

![image](https://user-images.githubusercontent.com/46308407/147901363-496cc331-fbb4-4231-a235-b4d4c0ec698f.png)

- It shows little correlation between the proposed memorization score and nearest neighbor based heuristics.
- NN test is unstable in at least two aspects: (i) memorization can happen locally, so NN comparison using subsets can fall short (ii) distance metric is non-trivial especially in high dimensional space.

## Comments

- Leave-One-Out based statistical approach for evaluation of deep learnings. (personally like the idea itself)
- Tackled interesting yet rather dim phenomenon of ***local memorization*** that is distinguished from commonly addressed overfitting.
- Relatively easy paper, both to write (presumably) and to read.
